Adding a tutorial for data driven root motion and root motion extract… (#1966)
o3de/o3de.org: content/docs/learning-guide/tutorials/animation/data-driven-root-motion.md

---
linkTitle: Root Motion
title: Data Driven Root Motion
description: Learn about data driven root motion, how to extract root motion from animation, and how to enable root motion on actors in Open 3D Engine (O3DE).
weight: 100
toc: true
---

*Root motion* is animation that moves an actor entity. This tutorial explains what root motion is in detail, and how to enable it for an actor.

For this tutorial, you'll need an actor asset with animation that has been processed by [Asset Processor](/docs/user-guide/assets/asset-processor/). The animation should move the actor off the scene origin. That is, the animation should move the actor in a direction, such as a walk cycle that moves the actor forward in space.

## What is root motion?

An actor entity usually has many components (meshes, a skeleton, colliders, and so on) that need to move together. If an actor jumps, the actor's collision capsule needs to have the same jump movement applied so that it moves with the skeleton and meshes. This movement, that moves all of the actor's components, is the *root motion*.

> Root motion can be *code driven* or *data driven*, The focus of this tutorial is data driven root motion, but it's helpful to understand both methods.

Reviewer: ```suggestion
Root motion can be *code driven* or *data driven*. The focus of this tutorial is data driven root motion, but it's helpful to understand both methods.
```

### Code driven

> For code driven root motion, the actor's animation cycles happen in place and any movement off the scene origin is provided by code (such as a script). The actor runs, walks, or jumps in place and the script provides the appropriate directional movement.

Reviewer: ```suggestion
For code driven root motion, the actor's animation cycles happen in place, and any movement off the scene origin is provided by code (such as a script). The actor runs, walks, or jumps in place, and the script provides the appropriate directional movement.
```

{{< /note >}}

### Which should you use?

Whether you use code driven or data driven root motion depends on your project's needs.

Code driven root motion might be preferable in scenarios where precise actor movement for play mechanics is more critical than visual performance. Platform games and fighting games, for example, usually require accurate actor movement. Ensuring an actor moves a precise distance in response to an input might be far more important than ensuring actor's movement relative to animation is visually accurate.

Data driven root motion can give animators more control over the actor's performance. Adventure games and first person games that strive for realism and immersion often require the actor's movement express some game condition or emotion. In these scenarios, the actor's movement relative to animation might prioritize visual accuracy.

> The remainder of this tutorial focuses on data driven root motion. You'll learn how to determine if an actor is properly set up for data driven root motion, how to extract root motion if necessary, and how to enable root motion in [Animation Editor](/docs/user-guide/visualization/animation/animation-editor/user-interface).

Reviewer: ```suggestion
The remainder of this tutorial focuses on data driven root motion. You'll learn how to determine if an actor is properly set up for data driven root motion, extract root motion if necessary, and enable root motion in [Animation Editor](/docs/user-guide/visualization/animation/animation-editor/user-interface).
```
You could drop the second and third occurrence of "how to" if you think it sounds better.

> ## Root motion set up

Reviewer: ```suggestion
## Root motion setup
```

The data in data driven root motion is usually keyframe animation applied to the root bone of an actor's skeleton by an animator. In an ideal scenario, the actor and animation source scene files follow these rules:

* The skeleton has a root bone placed at the scene origin.
* The root bone is connected to the actor's pelvis bone (or some relative bone on an actor that isn't a biped).
* Any motion that moves the actor off the scene origin is keyframed on the root bone.

> ### Check actor set up

Reviewer: ```suggestion
### Check actor setup
```

To determine if the actor has a root bone with animation, do the following:

1. In **O3DE Editor**, from the **Tools** menu, choose **Animation Editor**.

2. In Animation Editor, press **Ctrl + O** to open the actor. Select an actor from the file window.

3. Choose the {{< icon "visibility-on.svg" >}} **Visibility** button above the viewport, and select **Line skeleton** from the list to display the actor's skeleton as green lines.

4. In the **Motion Sets** tab, in the **Motion Set** group, choose the {{< icon "file-folder.svg" >}} **Folder** button to add a motion. Select a motion from the file window.

5. Choose the motion you just added to the Motion Set to enable it.

> 6. In the **Time View** tab, click the {{< icon "play.svg" >}} **Play** button to play the animation.

Reviewer 2: uber nit: It might be easier to just update all these to `1.` as the markdown viewer will the automatically update the numbers if you ever add anything in the middle in future for example, makes maintenance much easier 🙂

Author: Yes, I use the 1. notation when I write but there's some setting in my editor that keeps converting them.

Reviewer 2: Ah okay cool np :+1:

Author: fixed

| The actor moves off the scene origin, but one end of the green line that represents the root bone remains at the scene origin. | This actor doesn't have a root bone. The animation data is applied to another bone (such as the pelvis). You can either:<ul><li>add a root bone and animation to the source scene files in an animation application</li><li>attempt to [extract root motion](#root-motion-extraction) as explained in the next section</li></ul> |
| The actor moves off the scene origin, the root joint moves with the actor, and the actor snaps back to the scene origin when the animation cycle completes. | The actor has a root bone with animation data. You can move on to the [Enable root motion](#enable-root-motion) section. |

{{< important >}}
For the best result, edit the source scene files in your digital content creation application (Blender or Maya, for example) and add a root bone with animation.
{{< /important >}}

### Root motion extraction

> If your source scene files have animation that moves the actor off the scene origin, but the root bone doesn't move with the actor, it's likely that the source scene files don't have a root bone and that animation has bee applied to another bone such as the pelvis bone. You can try to extract the animation from the pelvis so it can be applied to the actor as root motion.

Reviewer 3: and that animation has BEEN

Author: good catch, fixed.

1. In Asset Browser, right-click the animation file and choose **Edit settings...** from the context menu.

1. In Scene Settings, select the **Motions** tab.

1. Choose **Add Modifier** and select **Root motion extraction** from the modifier list.

    {{< note >}}
A source scene file might contain more than one motion. You'll need to add a Root motion extraction modifier to each motion that requires it.
    {{< /note >}}

1. In the Root motion extraction modifier, ensure the **Sample joint** property is set to the bone containing the motion you want to extract as in the following example:

    {{< image-width src="/images/learning-guide/tutorials/animation/root-motion-extraction-modifier-bone.png" alt="Selecting the bone for root motion extraction in Scene Settings." >}}

1. By default, the bone's translation is extracted. You can choose to ignore the X or Y axis translation, and extract rotations by setting the properties in the modifier.

1. With the modifier configured, choose **Update** to save your custom processing rule. The animation is automatically processed by Asset Processor. Close the output window and exit Scene Settings.

> When the motion is processed, the root motion is extracted and applied to the root bone in the motion product asset. You can verify if the root motion extraction was successful by repeating the steps in the [Generate data driven root motion](#generate-data-driven-root-motion) section.

Author: ```suggestion
When the motion is processed, the root motion is extracted and applied to the root bone in the motion product asset. You can verify if the root motion extraction was successful by repeating the steps in the [Check actor setup](#check-actor-setup) section.
```

> When the motion is processed, the root motion is extracted and applied to the root bone in the motion product asset. You can verify if the root motion extraction was successful by repeating the steps in the [Generate data driven root motion](#generate-data-driven-root-motion) section.

Author: Good catch

The following table provides additional detail on the options provided by the Root motion extraction modifier:

| Option | Description |
| - | - |

> | **Sample Joint** | The bone that root motion is extracted from. Usually this is the hip or pelvis (or some relative bone for actors that aren't bipeds). |

Reviewer: ```suggestion
| **Sample Joint** | The bone that root motion is extracted from. Usually this is the hip or pelvis (or some relative bone for actors that aren't bipeds). |
```
```suggestion
| **Sample Joint** | The bone that root motion is extracted from. Usually, this is the hip or pelvis (or some relative bone for actors that aren't bipeds). |
```
